### DISCUSSION TITLE
How to do multi-step benchmarks?

### DISCUSSION THREAD

**Discussion by jeanplevesque** [dotnet/BenchmarkDotNet#2484] [Q&A] [answered]
Hi there, I'd like to create _multi-step_ benchmarks to measure each step of a longer routine.

Here's what I mean:
1. Instantiate an object.
2. Access object member A.
3. Access object member B.
4. Dispose the object.

I would like to benchmark this sequence and get the metrics (CPU and allocations) for each step (1, 2, 3, 4).
I've browsed the documentation for pointers but haven't found what I was looking for.

Is there a way to do something like this using BenchmarkDotNet?
Thanks!

**Comment by timcassell** [comment]
It's not simple (or portable), but you can do it with `[IterationSetup]` and `[IterationCleanup]`. You will need to store your object in an array, and do the step in a loop.

```cs
public class Benchmark
{
    // Adjust this value until BenchmarkDotNet is happy with benchmark time > 100ms
    const int invokeCount = 100_000;

    private MyObject[] _objects = new MyObject[invokeCount];

    [IterationSetup(Target = nameof(MemberA))]
    public void IterationSetup()
    {
        for (int i = 0; i < invokeCount; ++i)
        {
            _objects[i] = new MyObject();
        }
    }

    [IterationCleanup(Target = nameof(MemberA))]
    public void IterationSetup()
    {
        for (int i = 0; i < invokeCount; ++i)
        {
            _objects[i].Dispose();
        }
    }

    [Benchmark(OperationsPerInvoke = invokeCount)]
    public void MemberA()
    {
        for (int i = 0; i < invokeCount; ++i)
        {
            _objects[i].MemberA();
        }
    }
}
```

**Reply by jeanplevesque** [reply]
In my case I want to capture benchmarks for each step, not just the member access. From what I understand, `IterationSetup` and `IterationCleanup` don't capture benchmark data.

**Reply by timcassell** [reply, marked as answer]
My point was you have to create a separate benchmark for each step, each with its own setup and cleanup.
